omen2/object.py

```diff
@@ -0,0 +1,252 @@
+from .relation import Relation
+
+if TYPE_CHECKING:
+    from omen2.omen import Omen
+    from omen2 import Table
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class ObjMeta:
+    lock = RLock()
+    locked = False
+    new = True
+    table: Optional["Table"] = None
+    pk = None
+    restore = None
+
+
+# noinspection PyCallingNonCallable,PyProtectedMember
+class ObjBase:
+    # objects have 2 non-private attributes that can be overridden
+    _cascade = True
+    _pk: Tuple[str, ...] = ()  # list of field names in the db used as the primary key
+    _table_type: Type["Table"]  # class derived from Table
+
+    # objects should only have 1 variable in __dict__
+    _meta: ObjMeta = None
+
+    def __eq__(self, obj):
+        return obj._to_pk() == self._to_pk()
+
+    def __repr__(self):
+        return self.__class__.__name__ + "(" + str(self._to_pk()) + ")"
+
+    def __str__(self):
+        return str(self._to_db())
+
+    def __hash__(self):
+        return hash(self._to_pk_tuple())
+
+    def _to_pk_tuple(self):
+        return tuple(sorted(self._to_pk().items()))
+
+    def __lt__(self, other: "ObjBase"):
+        return self._to_pk_tuple() < other._to_pk_tuple()
+
+    def __init_subclass__(cls, **_kws):
+        # you must set these in the base class
+        assert cls._pk, "All classes must have a _pk"
+
+    def __init__(self, **kws):
+        """Override this to control initialization, generally calling it *after* you do your own init."""
+        self._meta = ObjMeta()
+        self._meta.lock = RLock()
+        self._check_kws(kws)
+        self._meta.new = True
+
+    def _save_pk(self):
+        self._meta.pk = self._to_pk()
+        self._meta.restore = self._to_db()
+
+    @classmethod
+    def _from_db(cls, dct):
+        """Override this if you want to change how deserialization works."""
+        return cls(**dct)
+
+    @classmethod
+    def _from_db_not_new(cls, dct):
+        """Override this if you want to change how deserialization works."""
+        ret = cls._from_db(dct)
+        ret._meta.new = False
+        ret._save_pk()
+        return ret
+
+    def _check_kws(self, dct):
+        for k in dct:
+            if k not in self.__dict__:
+                raise AttributeError(
+                    "%s not a known attribute of %s" % (k, self.__class__.__name__)
+                )
+
+    def _matches(self, dct):
+        for k, v in dct.items():
+            if getattr(self, k) != v:
+                return False
+        return True
+
+    def _update(self, dct):
+        self.__dict__.update(dct)
+
+    def _bind(self, table: "Table" = None, manager: "Omen" = None):
+        if table is None:
+            table = manager[self._table_type]
+        self._meta.table = table
+
+    def _to_db(self):
+        """Get dict of serialized data from self."""
+        ret = {}
+        for k in self._table_type.field_names:
+            v = getattr(self, k)
+            if v is None:
```

Comment:
I think this would prevent you from updating a value to `null`

Reply:
will write tests for that and fix, thanks